snuba/admin/clickhouse/migration_checks.py

```diff
@@ -0,0 +1,215 @@
+            if self.__migration_statuses[m]["status"] != Status.COMPLETED:
+                return ResultReason(False, Reason.NEEDS_EARLIER_MIGRATIONS)
+
+        return ResultReason(True, Reason.NO_REASON_NEEDED)
+
+    def can_reverse(self, migration_id: str) -> ResultReason:
+        """
+        Covers the following cases:
+        * no reversing a migration that is has not started
+        * no reversing a migration that has a subsequent migration that has not been run
+        """
+        all_migration_ids = self.all_migration_ids
+        if self.__migration_statuses[migration_id]["status"] == Status.NOT_STARTED:
+            return ResultReason(False, Reason.NOT_RUN_YET)
+
+        for m in all_migration_ids[all_migration_ids.index(migration_id) + 1 :]:
+            if self.__migration_statuses[m]["status"] != Status.NOT_STARTED:
+                return ResultReason(False, Reason.NEEDS_SUBSEQUENT_MIGRATIONS)
+
+        return ResultReason(True, Reason.NO_REASON_NEEDED)
+
+
+class PolicyChecker(Checker):
+    """
+    The PolicyChecker validates whether you can run or
+    reverse a migration based on the policy for the
+    migration's group.
+
+    Policies are defined in the ADMIN_ALLOWED_MIGRATION_GROUPS
+    setting.
+    """
+
+    def __init__(self, migration_group: MigrationGroup) -> None:
+        self.__migration_group: MigrationGroup = migration_group
+        self.__policy: MigrationPolicy = get_migration_group_polices()[
+            migration_group.value
+        ]
+
+    def _get_migration_key(self, migration_id: str) -> MigrationKey:
+        return MigrationKey(self.__migration_group, migration_id)
+
+    def can_run(self, migration_id: str) -> ResultReason:
+        key = self._get_migration_key(migration_id)
+        if self.__policy.can_run(key):
+            return ResultReason(True, Reason.NO_REASON_NEEDED)
+        else:
+            return ResultReason(False, Reason.RUN_POLICY)
+
+    def can_reverse(self, migration_id: str) -> ResultReason:
+        key = self._get_migration_key(migration_id)
+        if self.__policy.can_reverse(key):
+            return ResultReason(True, Reason.NO_REASON_NEEDED)
+        else:
+            return ResultReason(False, Reason.REVERSE_POLICY)
+
+
+def do_checks(
+    checkers: Sequence[Checker], migration_id: str
+) -> Tuple[ResultReason, ResultReason]:
+    """
+    Execute the can_run and can_reverse functionality
+    for the checkers.
+
+    Returns the failed ResultReason(s) for the first
+    check to fail in the sequence, otherwise returns
+    a passing ResultReason.
+    """
+
+    assert len(checkers) >= 1
+
+    for checker in checkers:
+        run_result = checker.can_run(migration_id)
+        if not run_result.allowed:
+            break
+
+    for checker in checkers:
+        reverse_result = checker.can_reverse(migration_id)
+        if not reverse_result.allowed:
+            break
+
+    return run_result, reverse_result
+
+
+def migration_checks_for_group(
```

Comment:
nit: if you want to make it clear what a method does, having a method name that does not contain a verb tends to be a bad sign (unless it returns a property). You probably want to rephrase it in a way that states what the method does (like `run_migrations....`.
Also this does not really run the checks for a group as it takes a list of migrations. Nothing says that it will run the checks for the entire group. It would run the checks on. the list of migrations you provide.